runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java

```diff
@@ -218,19 +220,55 @@ public TxnContext txn(@Nonnull final String namespace, IsolationLevel isolationL
     }
 
     /**
-     * Return the address of the latest updated made in this table.
+     * Return the address of the latest update made in this table.
+     *
+     * Note: we can't deliberately return the tail of the stream map
+     * as this entry might be a HOLE, we need to filter and return only that
+     * corresponding to the last DATA entry.
      *
      * @param namespace - namespace that this table belongs to.
      * @param tableName - table name of this table without the namespace prefixed in.
      * @return stream tail of this table.
      */
     public long getHighestSequence(@Nonnull final String namespace,
                                    @Nonnull final String tableName) {
-        return this.runtime.getSequencerView().query(
-                CorfuRuntime.getStreamID(
-                        TableRegistry.getFullyQualifiedTableName(namespace, tableName)
-                )
-        );
+
+        // Overall strategy:
+        // (1) Retrieve streams full address map from sequencer
+        // (2) Read entries in reverse to confirm it's a DATA type entry (for optimization
+        // read in batches). A batch size of 4 is guaranteed to suit one edge case case scenario: i.e.,
+        // consecutive checkpoints enforcing holes (3 rounds of CP accumulate before trim happens == 3 holes)
+
+        long startTime = System.nanoTime();
+        StreamAddressSpace streamAddressSpace = this.runtime.getSequencerView()
+                .getStreamAddressSpace(new StreamAddressRange(CorfuRuntime.getStreamID(
+                        TableRegistry.getFullyQualifiedTableName(namespace, tableName)), Address.MAX, Address.NON_ADDRESS));
+
+        int numBatches = 0;
+
+        if (streamAddressSpace.size() != 0) {
+            NavigableSet<Long> addresses = streamAddressSpace.copyAddressesToSet(Address.MAX).descendingSet();
+            Iterable<List<Long>> batches = Iterables.partition(addresses,
+                    runtime.getParameters().getHighestSequenceNumberBatchSize());
+
+            for (List<Long> batch : batches) {
+                numBatches++;
+                Map<Long, ILogData> entries = runtime.getAddressSpaceView().read(batch);
```

Comment:
Maybe a comment or TODO to indicate that we can optimize this later to not transfer actual data since we are only really interested in testing for holes?